### DISCUSSION TITLE
JDK 17 and GraalVM benchmarking

### DISCUSSION THREAD

**Discussion by ahmedeldeeb25-zz** [oracle/graal#4080] [Q&A] [answered]
Hello everybody,

I've started recently discovering **Graalvm, Micronauts,** and some new other tools.

I have watched many videos like this one https://www.youtube.com/watch?v=iXCVaQzXi5w&ab_channel=OracleDevelopers and did everything the same, but I didn't find any big difference in **Benchmarking** between Graalvm and JVM and I can't actually know what is the problem.

I'm using Graalvm 21 and java 17 to make the comparison.
I have tried many other examples and other versions of both JVM and Graal and have the same result.

I'm using **sdkman** to install and switch between different java versions.

**Note:** I have tried the same example in **Alina** talk, and some other examples and I can't find the big difference in Benchmarking between both.

**Comment by alina-yur** [comment, marked as answer]
Hi @ahmedeldeeb25, 

Thank you for asking this, I haven't run this example in a while. Just to clarify something here - in JIT mode, which we are talking about here, the GraalVM compiler actually runs in the context of a JVM. So, I ran [this app](https://github.com/graalvm/graalvm-demos/tree/master/java-simple-stream-benchmark) again with GraalVM Community 21.3 on my machine (which is not a proper benchmarking environment, but can give some basic understanding), and here are my results:
With the GraalVM compiler:
```
java -version
openjdk version "17.0.1" 2021-10-19
OpenJDK Runtime Environment GraalVM CE 21.3.0 (build 17.0.1+12-jvmci-21.3-b05)
OpenJDK 64-Bit Server VM GraalVM CE 21.3.0 (build 17.0.1+12-jvmci-21.3-b05, mixed mode, sharing)
java -jar target/benchmarks.jar 
...
Benchmark                             Mode  Cnt    Score    Error  Units
JavaSimpleStreamBenchmark.testMethod  avgt    3  156.553 ± 32.598  ns/op
```
Now disabling it:

```
java -jar -XX:-UseJVMCICompiler target/benchmarks.jar
...
Benchmark                             Mode  Cnt    Score    Error  Units
JavaSimpleStreamBenchmark.testMethod  avgt    3  233.679 ± 33.389  ns/op
```
which is about 33% speedup.
Of course the performance would depend on a lot of things, the actual workload being probably the most important one, but still I think this is a notable improvement. It’s also somewhat close to what we saw [recently reported](https://twitter.com/karlstoney/status/1456616777325158405) by the community.

**Reply by ahmedeldeeb25-zz** [reply]
Hi @alina-yur ,

Thank you for your help, I really appreciate your time.
Everything is obvious for me now, and it's time to prepare my presentation about GraalVM for my team.

**Comment by alina-yur** [comment]
@ahmedeldeeb25 you're welcome:) 

If you are interested in performance, particularly around startup, might also check out native image: https://www.graalvm.org/native-image/

For example this app is easy to start with and it shows clear benefits: https://guides.micronaut.io/latest/micronaut-creating-first-graal-app.html

Let me know if you need more help:)

Alina
